Prototype next gen i18n framework

bryk opened the pull request:
This is highly WIP with messy code, but I wanted to show it to look for
feedback.

Works on serve, serve:prod and build. Message extraction works too.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubernetes/dashboard/1239)
<!-- Reviewable:end -->

bryk commented:
@taimir @floreks Can you take a look and test?

bryk commented:
Yes, sorry for this. Install the deps manually if you can.

taimir commented:
Mhm, doing that just now.

bryk commented:
Run `gulp clean` & `gulp angular-templates` and see what is in .tmp dir. That's best way to test it.

taimir commented:
Nice! I can confirm that the file generation works. At a first glance, I don't see any major bugs in the code, either. I'll have more time for a thorough review later today or tomorrow morning :) Then we could consider alternative gulp libs (although none come to mind right now).

bryk commented:
Cool. Don't start full review until I post PTAL here. I need to clean up the code.

taimir commented:
Also one issue I just noticed that needs to be addressed: the string "xxx" has not been extracted to the `.xtb` files properly, even though it is in the generated `html.js` file.

bryk commented:
Ready for review. PTAL

@olekzabl Can you also review?

taimir commented:
First batch of comments, PTAL. Mostly about the regex.

---

Reviewed 2 of 8 files at r1, 6 of 9 files at r2.
Review status: 8 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[build/i18n.js, line 69 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRmxXe9osFZ7Fe6WEaz:-KRmxXe9osFZ7Fe6WEb-:b-14ixim) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L69)):*
> ```JavaScript
> 
> // Regex to match [[Foo | Bar]] i18n placeholders.
> const I18N_REGEX = /\[\[([^|]*?)(?:\|(.*?))?\]\]/mg;
> ```

Ok, to me that's the core of "core" of the whole exercise, so let's make sure it does what's intended:
* in the first capturing group there's `[^|]*?`, which is a lazy (as few as possible) capture of symbols different than `|`. Is there a reason to not leave it at the default "greedy" `[^|]`? It will always stop at `|` and wont continue further ... The only reason to make it "lazy" I can think of is if we allow the syntax to be both `[[ message | desription ]]` and `[[ message ]]`, omitting the description in the second. Then it would make sense to have it lazy, as otherwise it won't work.

* The second group, for the description, is *non-capturing* and *optional* at the same time:
`(?:\|(.*?))?`, where `?:` makes it non-capturing and the `?` at the end makes it optional. Because it's *non-capturing*, you wont an entry for it when you execute regex.exec(). That's fine. Now because it's *optional*, you are allowing for the syntax: `[[message]]`. But then there's an issue: the resulting `regex.exec()` for `[[message]]` will be :
```{js}
[ '[[message]]', 'message', undefined, index: 0, input: '[[message]]' ]
``` 
A bit below in the code, you are accessing `exec[2].trim()`, which obviously fails because exec[2] is undefined. Should be reproducible. This needs to be fixed.

* inside `(?:\|(.*?))?`, the lazy evalution of `(.*?)` is required so that the `]]` at the end are not consumed. This is correct, FWIW.

Please see if I understood your intentions correctly. If we are not gonna support `[[message]]` (without description) as proper syntax, this regex should be fixed.

---

*[build/i18n.js, line 87 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn58qH3MgnfcDaPz9f:-KRn58qH3MgnfcDaPz9g:bdvb29x) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L87)):*
> ```JavaScript
>     if (matches) {
>       return matches.map((match) => {
>         let exec = regexpClone(I18N_REGEX).exec(match);
> ```

Hmmm ... applying the regex twice seems a bit redundant, right? Or is there some magical performance boost when using regexpClone that I'm not aware of, that would be cool :D Anyways, I don't even know if it's possible to do it in another way. Will have to think about it.

---

*[build/i18n.js, line 88 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn5Y5qAauamTcO5QI_:-KRn5Y5qAauamTcO5QIa:bhs3chm) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L88)):*
> ```JavaScript
>       return matches.map((match) => {
>         let exec = regexpClone(I18N_REGEX).exec(match);
>         return {text: exec[1], desc: exec[2].trim(), original: match};
> ```

Here's where the bug described above happens.

---

*[build/i18n.js, line 100 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn6JMLFCKcbx-oZ7xd:-KRn6JMLFCKcbx-oZ7xe:b10wjh0) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L100)):*
> ```JavaScript
>    * @return {string}
>    */
>   function createMessageVarName(index) { return `MSG_${messageVarPrefix}_${index}`; }
> ```

I guess those functions are nested in `findI18nMessages()` for isolation reasons, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-:-KRn6uiJCqq-PGvATsDZ:bx8hs02)*
<!-- Sent from Reviewable.io -->

taimir commented:
Also some strange regex behavior on: https://regex101.com/

Namely, the following is not matched:

`[[very long  .............................................. message |
some description]]`

Whereas this one is matched:

`[[ very long ..............................................
message | some description]]`

Don't have the time to debug it right now, just wanted to let you know :)

---

Review status: 8 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-:-KRnJ00GQK_znFTRjiB_:beopr61)*
<!-- Sent from Reviewable.io -->

bryk commented:
> [[very long  .............................................. message |
> some description]]
Hmmm.... Tried this and all is matched. Take a look at what I've pushed here. 

PTAL. I've added some messages for debug purposes. 

---

Review status: 8 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[build/i18n.js, line 69 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRmxXe9osFZ7Fe6WEaz:-KTE-XCbTLdaoyENES7Y:b-m7wroz) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L69)):*
> in the first capturing group there's [^|]*?, which is a lazy (as few as possible) capture of symbols different than |. Is there a reason to not leave it at the default "greedy" [^|]? It will always stop at | and wont continue further ... The only reason to make it "lazy" I can think of is if we allow the syntax to be both [[ message | desription ]] and [[ message ]], omitting the description in the second. Then it would make sense to have it lazy, as otherwise it won't work.

That's correct. We need this to make description-less work. Should I add it to the comment? 

---

*[build/i18n.js, line 87 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn58qH3MgnfcDaPz9f:-KTE0jesOTYMgtAIzZdv:bqv00r8) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L87)):*
<details><summary><i>Previously, taimir (Atanas Mirchev) wrote…</i></summary>
> Hmmm ... applying the regex twice seems a bit redundant, right? Or is there some magical performance boost when using regexpClone that I'm not aware of, that would be cool :D Anyways, I don't even know if it's possible to do it in another way. Will have to think about it.
</details>
Regexes are stateful in JS... :D I.e., you cannot easily exec a regex twice. That's not trivial to know this, but yeah. That's why I'm copying this. 

---

*[build/i18n.js, line 88 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn5Y5qAauamTcO5QI_:-KTE1063FO8fn9qRZH9m:b-1zq2z1) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L88)):*
<details><summary><i>Previously, taimir (Atanas Mirchev) wrote…</i></summary>
> Here's where the bug described above happens.
</details>
Yep, fixed. 

---

*[build/i18n.js, line 100 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn6JMLFCKcbx-oZ7xd:-KTE2HSbDmPBOGAWfqgo:b5rrvrp) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L100)):*
<details><summary><i>Previously, taimir (Atanas Mirchev) wrote…</i></summary>
> I guess those functions are nested in `findI18nMessages()` for isolation reasons, right?
</details>
Yes. Nobody outside this function uses it, so why not isolate it? That was my thought process. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-:-KTE2aWFDBtLAv82LeQJ:b-2vj8z6)*
<!-- Sent from Reviewable.io -->

bryk commented:
PTAL

bryk commented:
@rf232 Can you also review this?

bryk commented:
@rf232 Can you also review this?

taimir commented:
Review status: 4 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[build/i18n.js, line 69 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRmxXe9osFZ7Fe6WEaz:-KTFErQxTIH8VHJoll4C:b-w89to9) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L69)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> > in the first capturing group there's [^|]*?, which is a lazy (as few as possible) capture of symbols different than |. Is there a reason to not leave it at the default "greedy" [^|]? It will always stop at | and wont continue further ... The only reason to make it "lazy" I can think of is if we allow the syntax to be both [[ message | desription ]] and [[ message ]], omitting the description in the second. Then it would make sense to have it lazy, as otherwise it won't work.
> 
> That's correct. We need this to make description-less work. Should I add it to the comment? 
</details>
I think that adding this could only be of benefit to future maintainers, just a short comment.

---

*[build/i18n.js, line 87 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn58qH3MgnfcDaPz9f:-KTFF4hR9DbaEeIg11wC:beg8hw2) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L87)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Regexes are stateful in JS... :D I.e., you cannot easily exec a regex twice. That's not trivial to know this, but yeah. That's why I'm copying this. 
</details>
Yeah, thought that was the case. It is ok, then.

---

*[build/i18n.js, line 88 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn5Y5qAauamTcO5QI_:-KTFFSpm6ZlxktHzCNpf:bfnfl0x) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L88)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Yep, fixed. 
</details>
I'll have a look at it later.

---

*[build/i18n.js, line 100 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn6JMLFCKcbx-oZ7xd:-KTFFKCMlnMTABpsfCiE:banqhqm) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L100)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Yes. Nobody outside this function uses it, so why not isolate it? That was my thought process. 
</details>
Yep, makes sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239)*
<!-- Sent from Reviewable.io -->

bryk commented:
PTAL. I've cleared this PR to be ready for merging. 

To test, fetch the branch and add messages to HTML files. 

---

Review status: 4 of 13 files reviewed at latest revision, 1 unresolved discussion.

---

*[build/i18n.js, line 69 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRmxXe9osFZ7Fe6WEaz:-KTIhFOhs8WaLx7x9ehu:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L69)):*
<details><summary><i>Previously, taimir (Atanas Mirchev) wrote…</i></summary>
> I think that adding this could only be of benefit to future maintainers, just a short comment.
</details>
Done.

---

*[build/i18n.js, line 88 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn5Y5qAauamTcO5QI_:-KTIhHzlRSs-EcbEXe-t:b-833310) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L88)):*
<details><summary><i>Previously, taimir (Atanas Mirchev) wrote…</i></summary>
> I'll have a look at it later.
</details>
PTAL 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-:-KTIifvBVug36NZoWJrp:b-og7nh3)*
<!-- Sent from Reviewable.io -->

taimir commented:
Reviewed 5 of 8 files at r3, 7 of 9 files at r4.
Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239)*
<!-- Sent from Reviewable.io -->

taimir commented:
Reviewed 1 of 9 files at r4.
Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239)*
<!-- Sent from Reviewable.io -->

taimir commented:
Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[build/i18n.js, line 88 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn5Y5qAauamTcO5QI_:-KTJ-6x8O466X7FW09jK:b-23m534) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L88)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> PTAL 
</details>
OK, finally some time for a review, I'll check it out now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239)*
<!-- Sent from Reviewable.io -->

taimir commented:
Reviewed 1 of 9 files at r2, 1 of 9 files at r4.
Review status: 4 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239)*
<!-- Sent from Reviewable.io -->

taimir commented:
Reviewed 1 of 8 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239)*
<!-- Sent from Reviewable.io -->

bryk commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[build/i18n.js, line 88 at r2](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KRn5Y5qAauamTcO5QI_:-KTJCW6IcENBKQTHGFds:b2iwc26) ([raw file](https://github.com/kubernetes/dashboard/blob/06ebe7a182673a07296f763e9ff6b8ffdff90b25/build/i18n.js#L88)):*
<details><summary><i>Previously, taimir (Atanas Mirchev) wrote…</i></summary>
> OK, finally some time for a review, I'll check it out now.
</details>
Is this LGMTed? Or are you still reviewing? Reviewable says it is green. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239)*
<!-- Sent from Reviewable.io -->

olekzabl commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[build/i18n.js, line 70 at r4](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KTJ97u9MKfkOv-DBbhP:-KTJ97uAmKFt_iSemNH2:brmhguu) ([raw file](https://github.com/kubernetes/dashboard/blob/d63256d59a0113a411ce69749810f240d43efe72/build/i18n.js#L70)):*
> ```JavaScript
> // Regex to match [[Foo | Bar]] or [[Foo]] i18n placeholders.
> // Technical details:
> // * First capturing group is lazy math for any string not-containing |. This is to make
> ```

math -> match?

---

*[build/i18n.js, line 74 at r4](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KTJETiIenukDJyBcPkA:-KTJETiIenukDJyBcPkB:bh3x75d) ([raw file](https://github.com/kubernetes/dashboard/blob/d63256d59a0113a411ce69749810f240d43efe72/build/i18n.js#L74)):*
> ```JavaScript
> // * Second is non-capturing and optional. It has a capturing group inside. This is to
> //   extract description that is optional.
> const I18N_REGEX = /\[\[([^|]*?)(?:\|(.*?))?\]\]/mg;
> ```

regex101.com suggests sensitivity to line breaks: 
https://screenshot.googleplex.com/WtENsyecqoi but https://screenshot.googleplex.com/XuHbdv9w0eZ.
But I don't see what could be wrong in the expression. Maybe that page is not good for our purposes?
Anyway, if the second text should be matched, can you verify that it indeed is?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239)*
<!-- Sent from Reviewable.io -->

taimir commented:
Excellent stuff, you can put variable bindings in the messages as was originally intended! I love this part :)

In any case, multi-line strings in the HTML templates get minified (if I'm not mistaken), before the regex is applied. So even though regex101 says they are not matched, this is not a big issue.

I'm ready to LGTM this once this has been cleared up.

taimir commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[build/i18n.js, line 74 at r4](https://reviewable.io:443/reviews/kubernetes/dashboard/1239#-KTJETiIenukDJyBcPkA:-KTJPbBZPutLZoZRymVj:b-urlw1j) ([raw file](https://github.com/kubernetes/dashboard/blob/d63256d59a0113a411ce69749810f240d43efe72/build/i18n.js#L74)):*
<details><summary><i>Previously, olekzabl wrote…</i></summary>
> regex101.com suggests sensitivity to line breaks: 
> https://screenshot.googleplex.com/WtENsyecqoi but https://screenshot.googleplex.com/XuHbdv9w0eZ.
> But I don't see what could be wrong in the expression. Maybe that page is not good for our purposes?
> Anyway, if the second text should be matched, can you verify that it indeed is?
</details>
Maybe it's an issue with the site, also see my last comment. If the HTML-minifier ditches the newlines prior to the message processing, I'd rather say lets merge this and maybe double check the regex later. It greatly benefits the development speed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1239)*
<!-- Sent from Reviewable.io -->

bryk commented:
Anyone can add review to GitHub to unblock merging? @olekzabl @taimir @rf232

taimir commented:
See [here](https://help.github.com/articles/about-pull-request-reviews/), section `Required reviews`:

>When required reviews are enabled, anyone with access to the repository can approve changes in a pull request. However, to merge your pull request you need someone who has write or admin permissions in the repository to approve your pull request's changes in their review.

bryk commented:
Oh, I see. @rf232 @olekzabl

rf232 commented on build/serve.js:
```diff
@@ -108,8 +108,8 @@ function serveDevelopmentMode() {
         conf.paths.serve,
-        conf.paths.frontendSrc,  // For angular templates to work.
-        conf.paths.app,          // For assets to work.
+        // conf.paths.frontendSrc,  // For angular templates to work.
```
Why leave this in in commented form?

bryk replied:
Done. PTAL

bryk commented:
PTAL
